x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/TypeResolutions.java

```diff
@@ -132,6 +132,28 @@ public static TypeResolution isFoldable(Expression e, String operationName, Para
         return TypeResolution.TYPE_RESOLVED;
     }
 
+    public static TypeResolution isNonNullFoldable(Expression e, String operationName, ParamOrdinal paramOrd) {
+        var foldableResolution = isFoldable(e, operationName, paramOrd);
+
+        if (foldableResolution.unresolved()) {
+            return foldableResolution;
+        }
+
+        if (e.fold() == null) {
+            return new TypeResolution(
+                format(
+                    null,
+                    "{}argument of [{}] can't be null, received [{}]",
+                    paramOrd == null || paramOrd == DEFAULT ? "" : paramOrd.name().toLowerCase(Locale.ROOT) + " ",
+                    operationName,
+                    Expressions.name(e)
+                )
+            );
+        }
+
+        return TypeResolution.TYPE_RESOLVED;
```

Comment:
```suggestion
        TypeResolution resolution = isFoldable(e, operationName, paramOrd);
        if (resolution.unresolved()) {
            return resolution;
        }

        if (e.dataType() == DataType.NULL || e.fold() == null) {
            resolution = new TypeResolution(
                format(
                    null,
                    "{}argument of [{}] cannot be null, received [{}]",
                    paramOrd == null || paramOrd == DEFAULT ? "" : paramOrd.name().toLowerCase(Locale.ROOT) + " ",
                    operationName,
                    Expressions.name(e)
                )
            );
        }

        return resolution;
```

@not-napoleon thanks for the ping 👍.
We have such types of checks (where we attempt to fold if it's foldable) to see if we can reason about that value for a validation.
The changes I am suggesting here:
- check also the `NULL` data type (`row` can generate a `NULL` data type on purpose). Search for similar usages of `.fold()` in esql and ql to get some inspiration on how this `.fold()` can be used.
- a code style that's usually predominant in the QL code area
- `can't` (more colloquial) -> `cannot` (more formal)